plugins/synapticsmst/synapticsmst-device.c

```diff
@@ -586,139 +1094,242 @@ synapticsmst_device_write_firmware (SynapticsMSTDevice *device,
-		return FALSE;
+	else{
+		code_size = (*(payload_data + offset) << 8) + *(payload_data + offset + 1);
+		if (code_size >= 0xFFFF) {
+			g_set_error_literal (error,
+						G_IO_ERROR,
+						G_IO_ERROR_INVALID_DATA,
+						"invalid firmware size");
+			return FALSE;
+		}
 	}
+	/* Firmware Checksum */
+	if(isChipPanamera == TRUE){
+		guint32 i =0;
+		for( i = 0; i < code_size + 1; i++ )
+		{
+			checksum += *(payload_data + offset + 0x10 + i);
+		}
+		if ((checksum & 0xFFFF)!=((*(payload_data + 0x408) << 8) | (*(payload_data + 0x409)))) {
+			g_set_error_literal (error,
+						G_IO_ERROR,
+						G_IO_ERROR_INVALID_DATA,
+						"firmware checksum error");
+			return FALSE;
+		}
+		checksum = 0;
+		offset = EEPROM_ESM_OFFSET;
+		for (i = 0; i < (payload_len - EEPROM_ESM_OFFSET); i++)
+		{
+			checksum += *(payload_data + offset + i);
+		}
+		if ((checksum & 0xFFFF)!=((*(payload_data + 0x40A) << 8) | (*(payload_data + 0x40B)))) {
+			g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+							"ESM firmware checksum error");
+			return FALSE;
+		}		
+	}
+	else{	
+		for (guint32 i = 0; i < (code_size + 17); i++)
+			checksum += *(payload_data + offset + i);
 
+		if (checksum & 0xFF) {
+			g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+						"firmware checksum error");
+			return FALSE;
+		}
+	}
 	/* TODO: May need a way to override this to cover field
 	 * issues of invalid firmware flashed*/
 	/* check firmware and board ID again */
 	tmp = (*(payload_data + ADDR_CUSTOMER_ID) << 8) + *(payload_data + ADDR_BOARD_ID);
-	if (tmp != synapticsmst_device_get_board_id (device)) {
-		g_set_error_literal (error,
-				     G_IO_ERROR,
-				     G_IO_ERROR_INVALID_DATA,
-				     "board ID mismatch");
-		return FALSE;
+	if(isChipPanamera == TRUE){
+		g_debug("bypass check CusID %x %x",tmp, synapticsmst_device_get_board_id (device));	
 	}
-
+	else {
+		if (tmp != synapticsmst_device_get_board_id (device)) {
+			g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+						"board ID mismatch");
+			return FALSE;
+		}
+	}
+	/* open device */
 	if (!synapticsmst_device_open (device, error)) {
 		g_prefix_error (error,
 				"can't open DP Aux node %s",
 				synapticsmst_device_get_aux_node (device));
 		return FALSE;
 	}
-
 	/* enable remote control */
 	if (!synapticsmst_device_enable_remote_control (device, error))
 		return FALSE;
 
-	/* erase SPI flash */
-	connection = synapticsmst_common_new (priv->fd, priv->layer, priv->rad);
-	if (synapticsmst_common_rc_set_command (connection,
-						UPDC_FLASH_ERASE,
-						2, 0, (guint8 *)&erase_code)) {
-		g_set_error_literal (error,
-				     G_IO_ERROR,
-				     G_IO_ERROR_INVALID_DATA,
-				     "can't erase flash");
-		return FALSE;
+	/* detect SPI quad mode & disable HDCP 2.2 */
+	if(isChipPanamera == TRUE)
+	{
+		/* Need to detect flash mode and ESM first ? */
+		/* disable flash Quad mode and ESM/HDCP2.2*/
+		for(guint8 i=0; i<16; i++){
+			dwData[i] = 0;	
```

Comment:
You only write length 4 for each of these.  So shouldn't you only need to zero out 4 bytes?